sql/src/test/java/io/crate/metadata/RoutingTest.java

```diff
@@ -105,4 +108,20 @@ public void testMixedShardAllocationsAreMerged() {
         assertThat(routing.locations().get("node-1").get("index-0"), containsInAnyOrder(3, 4, 6));
         assertThat(routing.locations().get("node-1").get("index-1"), containsInAnyOrder(3, 4, 6));
     }
+
+    @Test
+    public void testRoutingForRandomMasterOrDataNode() throws IOException {
+        Set<DiscoveryNode.Role> master_and_data = ImmutableSet.of(DiscoveryNode.Role.MASTER, DiscoveryNode.Role.DATA);
+        Map<String, String> attr = ImmutableMap.of();
+        DiscoveryNodes nodes = new DiscoveryNodes.Builder()
+            .add(new DiscoveryNode("data_master_node_1", LocalTransportAddress.buildUnique(), attr, master_and_data, null))
+            .add(new DiscoveryNode("data_master_node_2", LocalTransportAddress.buildUnique(), attr, master_and_data, null))
+            .add(new DiscoveryNode("client_node_1", LocalTransportAddress.buildUnique(), attr, ImmutableSet.of(), null))
+            .add(new DiscoveryNode("client_node_2", LocalTransportAddress.buildUnique(), attr, ImmutableSet.of(), null))
+            .add(new DiscoveryNode("client_node_3", LocalTransportAddress.buildUnique(), attr, ImmutableSet.of(), null))
+            .build();
+        ClusterState state = new ClusterState.Builder(new ClusterName("my_cluster")).nodes(nodes).build();
+        Routing routing = Routing.forRandomMasterOrDataNode(new TableIdent("doc", "table"), state);
```

Comment:
Maybe make sense to add a small for loop here to assert that client nodes are never returned.